driver-core/src/test/java/com/datastax/driver/core/UnresolvedUserTypeTest.java

```diff
@@ -0,0 +1,126 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.util.Collection;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.collect.Lists;
+import org.testng.annotations.Test;
+
+import com.datastax.driver.core.utils.CassandraVersion;
+
+import static com.datastax.driver.core.Assertions.assertThat;
+import static com.datastax.driver.core.DataType.cint;
+import static com.datastax.driver.core.DataType.list;
+import static com.datastax.driver.core.DataType.map;
+import static com.datastax.driver.core.DataType.set;
+
+@CassandraVersion(major = 3.0)
+public class UnresolvedUserTypeTest extends CCMBridge.PerClassSingleNodeCluster {
+
+    @Test(groups = "short")
+    public void should_resolve_nested_user_types() throws ExecutionException, InterruptedException {
+
+        Metadata metadata = cluster.getMetadata();
+        KeyspaceMetadata keyspaceMetadata = metadata.getKeyspace(this.keyspace);
+
+        UserType a = keyspaceMetadata.getUserType("\"A\"");
+        UserType b = keyspaceMetadata.getUserType("\"B\"");
+        UserType c = keyspaceMetadata.getUserType("\"C\"");
+        UserType d = keyspaceMetadata.getUserType("\"D\"");
+        UserType e = keyspaceMetadata.getUserType("\"E\"");
+        UserType f = keyspaceMetadata.getUserType("\"F\"");
+        UserType g = keyspaceMetadata.getUserType("g");
+        UserType h = keyspaceMetadata.getUserType("h");
+
+        assertThat(a).hasField("f1", c);
+        assertThat(b).hasField("f1", set(d));
+        assertThat(c).hasField("f1", map(e, d));
+        assertThat(d).hasField("f1", metadata.newTupleType(f, g, h));
+        assertThat(e).hasField("f1", list(g));
+        assertThat(f).hasField("f1", h);
+        assertThat(g).hasField("f1", cint());
+        assertThat(h).hasField("f1", cint());
+
+        // each CREATE TYPE statement has triggered a partial schema refresh,
+        // so nested UDTs are all unresolved
```

Comment:
I agree that it's how it works currently, but we should be able to do better than that. If `g` already exists in the metadata and we create `E` that references `g`, then a partial refresh that only targets `E` should not result in an unresolved type.
We could do this by allowing `buildUserTypes` to reference "old" types, only in the case where it's been triggered by a partial refresh.